### PR TITLE
WIP: use appropriate types for all resources instead of unstructured.Unstructured

### PR DESCRIPTION
**Pull request by rcernich**
Signed-off-by: rcernich <rcernich@redhat.com>

### CODE REVIEW

**Comment by rcernich** [comment]
Currently, untested...

**Comment by luksa** [comment]
@rcernich I've been working on getting this to run. My fixes are in https://github.com/luksa/istio-operator/commits/pr285-fixes

Some of these fixes are quick hacks/workarounds (especially the commit where I add the `ServiceMeshPolicy` type into the code instead of importing it from github.com/Maistra/istio - I'm not even sure why the type is declared there instead of in istio/api).

With my fixes, the operator now appears to successfully deploy the control plane (there are no errors in the log. I'm trying bookinfo now).

**Comment by dgn** [comment]
> Some of these fixes are quick hacks/workarounds (especially the commit where I add the `ServiceMeshPolicy` type into the code instead of importing it from github.com/Maistra/istio - I'm not even sure why the type is declared there instead of in istio/api).

That's because we introduced it, it's a replacement for the upstream `MeshPolicy` which is cluster-scoped.

**Comment by luksa** [comment]
> That's because we introduced it, it's a replacement for the upstream `MeshPolicy` which is cluster-scoped.

I know. But I'd expect `MeshPolicy` to be in istio/api instead of istio/istio and then `ServiceMeshPolicy` in our own fork of istio/api.

**Comment by dgn** [comment]
> I know. But I'd expect `MeshPolicy` to be in istio/api instead of istio/istio and then `ServiceMeshPolicy` in our own fork of istio/api.

I agree, that would probably make more sense. I'm hoping though that in the near future our API won't diverge anymore and this will be unnecessary

**Comment by rcernich** [comment]
Thanks @luksa for taking this the last mile.  Does it improve performance during reconcile at all, specifically prune?

Also, don't worry about the template fixes, as those are handled in MAISTRA-1062

File: pkg/controller/common/manifestprocessing.go [hunk 1]
```diff
@@ -61,8 +62,7 @@ func (p *ManifestProcessor) ProcessManifests(manifests []manifest.Manifest, comp
 				allErrors = append(allErrors, err)
 				continue
 			}
-			obj := &unstructured.Unstructured{}
-			_, _, err = unstructured.UnstructuredJSONScheme.Decode(rawJSON, nil, obj)
+			obj, err := runtime.Decode(p.JSONSerializer, rawJSON)
```
Comment (rcernich) [inline pkg/controller/common/manifestprocessing.go:65]: This should be changed to use 'unstructured.JSONFallbackEncoder'.  That should fix the issues with unregistered resources in the scheme.

Comment (rcernich) [reply pkg/controller/common/manifestprocessing.go:65]: Whoops, scratch that, that's only for Encoding.

File: pkg/controller/common/patch.go [hunk 2]
```diff
@@ -103,7 +107,7 @@ func (p *PatchFactory) CreatePatch(current, new runtime.Object) (Patch, error) {
 		if err != nil {
 			return nil, err
 		}
-		newObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, newBytes)
+		newObj, err := runtime.Decode(p.serializer, newBytes)
```
Comment (rcernich) [inline pkg/controller/common/patch.go:110]: This needs to fallback to unstructured if the type is unknown.

File: pkg/controller/common/patch.go [hunk 3]
```diff
@@ -130,7 +134,7 @@ func (p *PatchFactory) CreatePatch(current, new runtime.Object) (Patch, error) {
...
```
Comment (rcernich) [inline pkg/controller/common/patch.go:137]: Ditto, fallback to unstructured...

File: pkg/controller/common/patch.go (hunk 2 above)
Comment (rcernich) [reply pkg/controller/common/patch.go:110]: Line 83 should use our `scheme`, not the generic `runtime.Scheme`.

**Comment by luksa** [comment]
A quick performance test shows that these changes speed up the operator by 3 to 4 times. 

Time spent in controller's Reconcile() function:

**Unstructured**
1st create: 	68s (prune 14s)   
1st delete: 	32s (prune 30s)
2nd create:	57s 
2nd delete:	32s 

**Strongly-typed**
1st create: 	25s (prune 4s)
1st delete: 	10s (prune 9s)
2nd create:	13s
2nd delete:	10s

**Comment by rcernich** [comment]
FYI @ostromart

**Comment by rcernich** [comment]
We might be able to eek out a bit more by integrating the client-go library for istio.  At that point, I think there will only be one or two resource types that still use unstructured.

**Comment by dgn** [comment]
Should we open a new PR from @luksa's branch?

**Comment by rcernich** [comment]
That sounds good to me, although I suspect there will be some issues with the rebase, given all the recent refactoring. :(

**Comment by jwendell** [comment]
@rcernich What's the status of this? Obsolete and safe to close? It targets 1.1.

**Comment by luksa** [comment]
Obsolete. We did most (if not all) of this in other PRs. Closing.

**Pull request closed without merging**